reference/plugins.md

```diff
@@ -2,7 +2,170 @@
+  # let's include Singleton.
+  include Singleton
+
+  # Set default options, in this case the assumed path to the 
+  # configuration file (relative to the application's root)
+  def initialize
+    @options = {config_file: 'config/authentication.yml'}
+  end
+
+  # Hardcode (because it's a singleton) where Praxis should
+  # add the plugin's configuration definition to the application's.
+  def config_key
+    :authentication
+  end
+
+  # Define the plugin's configuration attribute(s), which in this case
+  # is just a simple boolean for whether to require authentication for actions
+  # by default, or only only specific actions.
+  def prepare_config!(node)
+    node.attributes do
+      attribute :authentication_default, Attributor::Boolean, default: false,
+        description: 'Require authentication for all actions?'
+    end
+  end
+
+  # Implement a simple authenticate method that does nothing useful
+  # other than illustrate doing something.
+  def authenticate(request)
+    request.current_user == 'guest'
+  end
+
+end
+{% endhighlight %}
+
+The plugin's `Request` module will be included in to `Praxis::Request`. So we add a simple `current_user` method.
+{% highlight ruby %}
+module Request
+  def current_user
+    'guest'
+  end
+end
+{% endhighlight %}
+
+Extensions to `Praxis::Controller` will typically want to register `before`, `after`, or `around` callbacks. Ours is no different:
+{% highlight ruby %}
+module Controller
+  # extend AS::Concern so that we can use its included callback to easily register our callback in concrete controllers. Note that Praxis::Controller
+  # also extends AS::Concern.
+  extend ActiveSupport::Concern
+
+  # and register our before :action in the included callbak
+  # from AS::Concern
+  included do
+    before :action do |controller|
+      action = controller.request.action
+      if action.authentication_required
+        unless Plugin.authenticate(controller.request)
+          return Praxis::Responses::Unauthorized.new(body: 'unauthorized')
+        end
+      end
+    end
+  end
+
+end
+{% endhighlight %}
+
+We add a handy `requires_authentication` method to action definitions in the `ActionDefinition` module.
+{% highlight ruby %}
+module ActionDefinition
+  # and again, extend AS::Concern. This time so we can use
+  # decorate_docs in the proper context.
+  extend ActiveSupport::Concern
+
+
+  included do
+    # add an :authentication_required key to the generated documentation.
+    # note: this is only for the generated JSON output, handling the display
+    # in the doc browser is separate.
+    decorate_docs do |action, docs|
+      docs[:authentication_required] = action.authentication_required
+    end
+  end
+
+  # simple requires_authentication dsl helper.
+  def requires_authentication(value)
+    @authentication_required = value
+  end
+
+  # and a quick authentication_required getter, defaulting to false.
+  def authentication_required
+    @authentication_required ||= false
```

Comment:
Is there a possibility that someone could call authentication_required before requires_authentication(true) and get false returned instead of true?